File: Listacirenlazada/Listacirenlazada.java
public class Nodo<T> {
    public T data;
    public Nodo<T> next;

    public Nodo(T data) {
        this.data = data;
        this.next = null;
    }
}

public class ListaCircularEnlazada<T> {
    private Nodo<T> head;

    public ListaCircularEnlazada() {
        this.head = null;
    }

    public void insertarAlPrincipio(T data) {
        Nodo<T> nuevoNodo = new Nodo<>(data);
        if (this.head == null) {
            this.head = nuevoNodo;
            nuevoNodo.next = this.head;
        } else {
            Nodo<T> nodoActual = this.head;
            while (nodoActual.next != this.head) {
                nodoActual = nodoActual.next;
            }
            nuevoNodo.next = this.head;
            this.head = nuevoNodo;
            nodoActual.next = this.head;
        }
    }

    public void insertarAlFinal(T data) {
        Nodo<T> nuevoNodo = new Nodo<>(data);
        if (this.head == null) {
            this.head = nuevoNodo;
            nuevoNodo.next = this.head;
        } else {
            Nodo<T> nodoActual = this.head;
            while (nodoActual.next != this.head) {
                nodoActual = nodoActual.next;
            }
            nodoActual.next = nuevoNodo;
            nuevoNodo.next = this.head;
        }
    }

    public void imprimirLista() {
        if (this.head == null) {
            System.out.println("None");
            return;
        }

        Nodo<T> nodoActual = this.head;
        do {
            System.out.print(nodoActual.data + " -> ");
            nodoActual = nodoActual.next;
        } while (nodoActual != this.head);
        System.out.println("(vuelve a " + this.head.data + ")");
    }

    public boolean buscar(T dataBuscada) {
        if (this.head == null) {
            return false;
        }

        Nodo<T> nodoActual = this.head;
        do {
            if (nodoActual.data.equals(dataBuscada)) {
                return true;
            }
            nodoActual = nodoActual.next;
        } while (nodoActual != this.head);
        return false;
    }

    public void eliminar(T dataAEliminar) {
        if (this.head == null) {
            return;
        }

        if (this.head.data.equals(dataAEliminar)) {
            if (this.head.next == this.head) {
                this.head = null;
                return;
            }

            Nodo<T> nodoActual = this.head;
            while (nodoActual.next != this.head) {
                nodoActual = nodoActual.next;
            }

            nodoActual.next = this.head.next;
            this.head = this.head.next;
            return;
        }

        Nodo<T> nodoPrevio = this.head;
        Nodo<T> nodoActual = this.head.next;
        while (nodoActual != this.head) {
            if (nodoActual.data.equals(dataAEliminar)) {
                nodoPrevio.next = nodoActual.next;
                return;
            }
            nodoPrevio = nodoActual;
            nodoActual = nodoActual.next;
        }
    }

    public static void main(String[] args) {
        ListaCircularEnlazada<Integer> miLista = new ListaCircularEnlazada<>();
        miLista.insertarAlFinal(10);
        miLista.insertarAlFinal(20);
        miLista.insertarAlFinal(30);
        miLista.imprimirLista();

        miLista.insertarAlPrincipio(5);
        miLista.imprimirLista();

        System.out.println("¿Está el 20? " + miLista.buscar(20));
        System.out.println("¿Está el 99? " + miLista.buscar(99));

        miLista.eliminar(20);
        miLista.imprimirLista();

        miLista.eliminar(5);
        miLista.imprimirLista();

        miLista.eliminar(30);
        miLista.imprimirLista();
    }
}

File: Listacirenlazada/Listacirenlazada.cs
using System;
using System.Collections.Generic;

public class Nodo<T>
{
    public T Data { get; set; }
    public Nodo<T> Next { get; set; }

    public Nodo(T data)
    {
        this.Data = data;
        this.Next = null;
    }
}

public class ListaCircularEnlazada<T>
{
    private Nodo<T> head;

    public ListaCircularEnlazada()
    {
        this.head = null;
    }

    public void InsertarAlPrincipio(T data)
    {
        Nodo<T> nuevoNodo = new Nodo<T>(data);
        if (this.head == null)
        {
            this.head = nuevoNodo;
            nuevoNodo.Next = this.head;
        }
        else
        {
            // Encontrar el último nodo para actualizar su puntero
            Nodo<T> tail = this.head;
            while (tail.Next != this.head)
            {
                tail = tail.Next;
            }
            nuevoNodo.Next = this.head;
            this.head = nuevoNodo;
            tail.Next = this.head; // El último nodo ahora apunta al nuevo head
        }
    }

    public void InsertarAlFinal(T data)
    {
        Nodo<T> nuevoNodo = new Nodo<T>(data);
        if (this.head == null)
        {
            this.head = nuevoNodo;
            nuevoNodo.Next = this.head;
        }
        else
        {
            Nodo<T> nodoActual = this.head;
            while (nodoActual.Next != this.head)
            {
                nodoActual = nodoActual.Next;
            }
            nodoActual.Next = nuevoNodo;
            nuevoNodo.Next = this.head;
        }
    }

    public void ImprimirLista()
    {
        if (this.head == null)
        {
            Console.WriteLine("None");
            return;
        }

        Nodo<T> nodoActual = this.head;
        do
        {
            Console.Write($"{nodoActual.Data} -> ");
            nodoActual = nodoActual.Next;
        } while (nodoActual != this.head);
        Console.WriteLine($"(vuelve a {this.head.Data})");
    }

    public bool Buscar(T dataBuscada)
    {
        if (this.head == null)
        {
            return false;
        }

        Nodo<T> nodoActual = this.head;
        do
        {
            if (EqualityComparer<T>.Default.Equals(nodoActual.Data, dataBuscada))
            {
                return true;
            }
            nodoActual = nodoActual.Next;
        } while (nodoActual != this.head);
        return false;
    }

    public void Eliminar(T dataAEliminar)
    {
        if (this.head == null)
        {
            return;
        }

        // Encontrar el nodo a eliminar y su predecesor
        Nodo<T> nodoPrevio = null;
        Nodo<T> nodoActual = this.head;

        do
        {
            if (EqualityComparer<T>.Default.Equals(nodoActual.Data, dataAEliminar))
            {
                break; // Nodo encontrado
            }
            nodoPrevio = nodoActual;
            nodoActual = nodoActual.Next;
        } while (nodoActual != this.head);

        // Si el nodo no se encontró (volvimos al inicio)
        if (!EqualityComparer<T>.Default.Equals(nodoActual.Data, dataAEliminar))
        {
            return; // El elemento no está en la lista
        }

        // Caso 1: La lista tiene un solo nodo
        if (nodoActual.Next == nodoActual)
        {
            this.head = null;
        }
        // Caso 2: El nodo a eliminar es el head
        else if (nodoActual == this.head)
        {
            nodoPrevio.Next = this.head.Next; // El último nodo apunta al nuevo head
            this.head = this.head.Next;
        }
        // Caso 3: El nodo está en medio o al final
        else
        {
            nodoPrevio.Next = nodoActual.Next;
        }
    }
}

public class Lisciren
{
    public static void Main(string[] args)
    {
        ListaCircularEnlazada<int> miLista = new ListaCircularEnlazada<int>();
        miLista.InsertarAlFinal(10);
        miLista.InsertarAlFinal(20);
        miLista.InsertarAlFinal(30);
        miLista.ImprimirLista();

        miLista.InsertarAlPrincipio(5);
        miLista.ImprimirLista();

        Console.WriteLine($"¿Está el 20? {miLista.Buscar(20)}");
        Console.WriteLine($"¿Está el 99? {miLista.Buscar(99)}");

        miLista.Eliminar(20);
        miLista.ImprimirLista();

        miLista.Eliminar(5);
        miLista.ImprimirLista();

        miLista.Eliminar(30);
        miLista.ImprimirLista();
    }
}